File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValid(root,low,high):
            if root is None:
                return True
            if root.val < high and root.val > low:
                a1 = isValid(root.left,low,root.val)
                a2 = isValid(root.right,root.val,high)
                if a1 and a2:
                    return True
                return False
            else:
                return False
        return isValid(root,-inf,inf)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long int num=LONG_MIN;
    bool isValid(TreeNode* root,long maxi,long mini)
    {
        if(!root) return true;
        if(mini<root->val and maxi>root->val)
        {
            return isValid(root->left,root->val,mini) && isValid(root->right,maxi,root->val);
        }
        return false;
    }
    bool isValidBST(TreeNode* root) {
        /*if(!root) return true;
        bool ans=isValidBST(root->left);
        if(root->val>num) num=root->val;
        else return false;

        return ans && isValidBST(root->right);*/

        bool ans=isValid(root,LONG_MAX,LONG_MIN);
        return ans;
        
    }
};